### PR TITLE
terraform-cloud-agent: initialize derivation & NixOS service module

### PR DESCRIPTION
**Pull request by ixmatus**
This change adds a derivation for the [`terraform-cloud-agent`][1] executable and a NixOS module that defines a systemd service for running that agent.

[1]: https://developer.hashicorp.com/terraform/cloud-docs/agents.

### CODE REVIEW

**Comment by ixmatus** [comment]
I'm opening the pull request so I can checkpoint the team programming work we've done on this. However, before I merge it (assuming I have approvals), I need to:

- Specify a specific user and group and drop privileges for the systemd service; and,
- Verify the NixOS module works.

@jsoo1 additionally, what release branch in our fork should I backport to?

**Comment by jsoo1** [comment]
No need to backport! Thank you!

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (jsoo1) [inline nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:41]: I recommend using a submodule for all the flags with a freeformType in addition to the known ones.

This is to make the module more future-proof to cli changes.
```suggestion
    tokenPath = lib.mkOption {
      type = lib.types.path;
      default = "/etc/terraform-cloud-agent-token";
    };

    data-dir = lib.mkOption {
      type = lib.types.nullOr lib.types.path;
      default = null;
    };

    cache-dir = lib.mkOption {
      type = lib.types.nullOr lib.types.path;
      default = null;
    };

    flags = lib.mkOption {
      freeformType = lib.types.attrsOf lib.types.str;

      name = lib.mkOption {
        type = lib.types.str;
      };

      log-level = lib.mkOption {
        type = lib.types.nullOr (lib.types.enum [ "trace" "debug" "info" "warn" "error" ]);
        default = null;
      };

      address = lib.mkOption {
        type = lib.types.nullOr lib.types.str;
        default = null;
      };

      accept = lib.mkOption {
        type = lib.types.commas;
        default = "plan,apply,policy,assessment";
      };

      auto-update = lib.mkOption {
        type = lib.types.str;
        readOnly = true;
        default = "disabled";
      };
    };
```

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (jsoo1) [inline nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:62]: I don't think there will ever be more than one subdirectory/be configurable (unless you mean to make a submodule for each `name`), thoughts?
```suggestion
      StateDirectory = "terraform-cloud-agent";
      CacheDirectory = "terraform-cloud-agent";
```

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (jsoo1) [inline nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:63]: This will ensure the directory in /etc/terraform-cloud-agent
```suggestion
      ConfigurationDirectory = "terraform-cloud-agent";
    };
```

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (jsoo1) [inline nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:35]: You can refer to the systemd configured directories via [specifiers](https://www.man7.org/linux/man-pages/man5/systemd.unit.5.html#SPECIFIERS):
```suggestion
      default = "%E/token";
```

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (jsoo1) [inline nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:52]: Ditto the [specifiers](https://www.man7.org/linux/man-pages/man5/systemd.unit.5.html#SPECIFIERS) here - they can be referred to in environment variables and will be pasted in:
```suggestion
    environment = {
      TFC_AGENT_DATA_DIR = "%S";
      TFC_AGENT_CACHE_DIR = "%C";
    };
    
    script = ''
      export TFC_AGENT_TOKEN=$(cat ${cfg.tokenPath})
```

**Review by jsoo1 — CHANGES_REQUESTED** [review]
Thank you! Looking like a solid first draft and I just have a couple overarching questions.  The only real blocker here is to use the freeform submodule and the systemd specifiers.

Question: What is the purpose of the `name` flag? Can there be multiple terraform-cloud-agents running at a time? For what end? If so maybe that's a reason to have multiple services, one per `name`?

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (ixmatus) [reply nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:54]: To use `getExe` I also need to add `meta.mainProgram` to the derivation to specify the main program.

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (ixmatus) [reply nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:63]: I don't think the tfc-agent will ever use a configuration file in etc though...

Comment (ixmatus) [reply nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:63]: Oh, nevermind, I see you using it for the tfc agent token file.

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (ixmatus) [reply nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:52]: Addressed in 0ffe3d1a6d69.

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (ixmatus) [reply nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:35]: Addressed in 63d4c4d6d1c5.

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (ixmatus) [reply nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:63]: Addressed in 3e1c32fefdf5.

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (ixmatus) [reply nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:62]: Addressed in 566796a2db62.

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (ixmatus) [reply nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:41]: TIL about freeformType. Thanks!

Addressed in 3d7a6140c9b4.

File: nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix
Comment (ixmatus) [reply nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix:54]: Addressed in 259c26b5e265 and e3121afcde0f.

File: pkgs/tools/infrastructure/terraform-cloud-agent/default.nix
Comment (ixmatus) [reply pkgs/tools/infrastructure/terraform-cloud-agent/default.nix:14]: Resolved in da98f2167a99.

**Comment by ixmatus** [comment]
Closing in favor of #203. @jsoo1 do you mind giving me a review of that PR?

**Pull request closed without merging**
